src/main/scala/org/constellation/rewards/RewardsManager.scala

```diff
@@ -0,0 +1,201 @@
+package org.constellation.rewards
+
+import cats.effect._
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import org.constellation.checkpoint.CheckpointService
+import org.constellation.consensus.Snapshot
+import org.constellation.domain.observation.Observation
+import org.constellation.p2p.PeerNotification
+import org.constellation.primitives.Schema.{AddressCacheData, CheckpointEdge}
+import org.constellation.primitives.{ChannelMessage, Transaction}
+import org.constellation.rewards.RewardsManager.rewardDuringEpoch
+import org.constellation.storage.{AddressService, SnapshotBroadcastService}
+import org.constellation.trust.TrustEdge
+
+class RewardsManager[F[_]: Concurrent](
+  eigenTrust: EigenTrust[F],
+  checkpointService: CheckpointService[F],
+  addressService: AddressService[F],
+  snapshotBroadcastService: SnapshotBroadcastService[F],
+) {
+  final val rewards: Ref[F, Map[String, Double]] = Ref.unsafe(Map.empty)
+
+  def updateBySnapshot(snapshot: Snapshot): F[Unit] =
+    for {
+      observations <- observationsFromSnapshot(snapshot)
+      _ <- eigenTrust.retrain(observations)
+
+      snapshotNum <- getSnapshotNumber
+
+      hashesTrustMap <- eigenTrust.getTrustForAddressHashes
+
+      distribution = RewardsManager
+        .rewardDistribution(hashesTrustMap.keys.toSeq, hashesTrustMap)
+        .mapValues(_ * rewardDuringEpoch(snapshotNum))
+
+      _ <- updateRewards(distribution)
+      _ <- updateAddressBalances(distribution)
+    } yield ()
+
+  // TODO: Check if this is the correct way of checking the snapshot number
+  // Assumption that `getRecentSnapshots` returns all snapshots may be failure
```

Comment:
Not sure if epoch operates on total number of snapshots or just on height (especially maxHeight - redownloadInterval). Because `curShapshot: Int` may have many meanings :)

Reply:
So what we want here is snapshot height, of snapshots that have been accepted past the height interval